platform/view/services/comm/p2p.go

```diff
@@ -242,24 +243,29 @@ func (s *streamHandler) handleIncoming() {
 		msg := &ViewPacket{}
 		err := s.reader.ReadMsg(msg)
 		if err != nil {
-			s.node.streamsMutex.Lock()
 			if s.node.isStopping {
 				if logger.IsEnabledFor(zapcore.DebugLevel) {
 					logger.Debugf("error reading message while closing. ignoring.", err.Error())
 				}
 				break
 			}
-			logger.Errorf("caught error: [%s][%s]", err.Error(), debug.Stack())
-			defer s.node.streamsMutex.Unlock()
 
-			for i, thisSH := range s.node.streams[s.stream.Conn().RemotePeer()] {
+			logger.Debugf("error reading message: [%s][%s]", err.Error(), debug.Stack())
+
+			// remove stream handler
+			remotePeerID := s.stream.Conn().RemotePeer()
+			s.node.streamsMutex.Lock()
+			for i, thisSH := range s.node.streams[remotePeerID] {
 				if thisSH == s {
-					s.node.streams[s.stream.Conn().RemotePeer()] = append(s.node.streams[s.stream.Conn().RemotePeer()][:i], s.node.streams[s.stream.Conn().RemotePeer()][i+1:]...)
+					s.node.streams[remotePeerID] = append(s.node.streams[remotePeerID][:i], s.node.streams[remotePeerID][i+1:]...)
 					s.wg.Done()
+					s.node.streamsMutex.Unlock()
 					return
 				}
 			}
+			s.node.streamsMutex.Unlock()
```

Comment:
this unlock is not invoked is line 263 (the return) is reached, no?

Reply:
correct, if we return before, this line is not reached. At this point the Unlock is kind of pointless as we panic anyway.

However, as @mffrench proposed to reduce panics whenever possible, the `panic` should replaces with better conflict resolution.

Reply:
Sorry, my fault. I didn't see the lock before return. All good.